src/Pact/Analyze/Eval/Term.hs

```diff
@@ -631,6 +636,135 @@ evalTerm = \case
       Some (SObject _) _ -> throwErrorNoLoc "We can't yet analyze calls to `hash` on objects"
       Some _ _           -> throwErrorNoLoc "We can't yet analyze calls to `hash` on non-{string,integer,bool}"
 
+  Pact steps -> local (inPact .~ sTrue) $ do
+    -- TODO: pending #442
+    _previouslyYielded <- use $ latticeState . lasYielded
+
+    -- We execute through all the steps once (via a left fold), then we execute
+    -- all the rollbacks (via for), in reverse order.
+
+    (rollbacks, _) <- ifoldlM
+      (\stepNo (rollbacks, successChoice)
+        --          ^ do we make it to this step?
+        (Step (tm :< ty) successPath {- s_n -} mEntity mCancelVid mRollback) ->
+        withSing ty $ do
+
+        -- The first step has no cancel var. All other steps do.
+        cancel <- case mCancelVid of
+
+          -- We can never cancel before executing the first step
+          Nothing -> pure sFalse
+
+          -- ... otherwise, we nondeterministically either succeed or cancel
+          Just (cancelPath {- c_n -}, cancelVid) -> withReset stepNo $ do
+            cancel <- view (aeNondets . at cancelVid)
+              ??? "couldn't find cancel var"
+
+            tagFork successPath cancelPath (sansProv successChoice)
+              (sansProv $ sNot cancel)
+
+            pure $ successChoice .&& cancel
+
+        let -- Trigger the latest rollback if we cancel on this step
+            rollbacks' = rollbacks & _head . _2 %~ (.|| cancel)
+            rollbacks'' = case mRollback of
+              Nothing
+                -> rollbacks'
+              Just (rollbackPath {- r_n -}, rollback)
+                -> (rollbackPath, sFalse, rollback):rollbacks'
+                --                ^ should we run this rollback, and every
+                --                  rollback before it?
+
+            reachesStep = successChoice .&& sNot cancel
+
+        ite reachesStep
+          (void $ evalTermWithEntity mEntity tm)
+          (pure ())
+
+        pure (rollbacks'', reachesStep))
+      ([], sTrue)
+      steps
+
+    void $ foldlM
+      (\alreadyRollingBack
+        (path {- r_n -}, rollbackTriggered, rollback) -> do
+        -- If this rollback was triggered, we execute this and all earlier
+        -- rollbacks
+        let nowRollingBack = alreadyRollingBack .|| rollbackTriggered
+        tagSubpathStart path $ sansProv nowRollingBack
+        ite nowRollingBack
+          (void $ evalETerm rollback)
+          (pure ())
+        pure nowRollingBack)
+      sFalse
+      rollbacks
+
+    pure "pact done"
```

Comment:
is this definitely what we want here? i guess it doesn't matter?

Reply:
Yeah I think it doesn't matter but I can change this to be a little more clear.